File: javascript/leetcode/first-unique-character-in-a-string.js
/**
 * @param {string} s
 * @return {number}
 */
var firstUniqChar = function(s) {
    var seen = {}
    for (var i = 0; i < s.length; i++) {
        if (seen[s[i]]) { seen[s[i]].count++ }
        else {seen[s[i]] = {firstIdx: i, count: 1}}
    }
    var earliestIdx = -1
    for (key in seen) {
        if (seen[key].count == 1) {
            if (earliestIdx == -1) {earliestIdx = seen[key].firstIdx}
            else if (seen[key].firstIdx < earliestIdx) {
                earliestIdx = seen[key].firstIdx
            }
        }
    }
    return earliestIdx
};

File: python/leetcode/first-unique-character-in-a-string.py
class Solution:
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: int
        """
        seen = {}
        for i in range(len(s)):
            if s[i] in seen:
                seen[s[i]]['count'] += 1
            else:
                seen[s[i]] = {'first_index': i, 'count': 1}
        earliest_index = -1
        for key in seen:
            if seen[key]['count'] == 1:
                if earliest_index == -1:
                    earliest_index = seen[key]['first_index']
                elif seen[key]['first_index'] < earliest_index:
                    earliest_index = seen[key]['first_index']
        return earliest_index
